File: LeetCode/687_LongestUnivaluePath.py
# https://leetcode.cn/problems/longest-univalue-path/
# 2022年09月02日 07:46:02

# v1 recursive DFS
class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        currMax=0
        def recursiveDFS(root):
            if not root : return 0
            leftMax=recursiveDFS(root.left)
            righMax=recursiveDFS(root.right)
            if root.left and root.val==root.left.val:
                leftMax+=1
            else:
                leftMax=0
            if root.right and root.val==root.right.val:
                righMax+=1
            else:
                righMax=0
            nonlocal currMax
            currMax=max(currMax,leftMax+righMax)
            return max(leftMax,righMax)
        recursiveDFS(root)
        return currMax

# v1.1 recursive without local variable
    def longestUnivaluePath(self, root):
        def recDFS2(root):
            if not root: return 0, 0
            maxL, pathL  = recDFS2(root.left)
            maxR, pathR = recDFS2(root.right)        
            pathL = pathL + 1 if root.left and root.left.val == root.val else 0
            pathR = pathR+ 1 if root.right and root.right.val == root.val else 0
            return max(maxL, maxR, pathL + pathR), max(pathL, pathR)
        return recDFS2(root)[0]



# v2 iterative post-order DFS
class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        currMax = 0
        postorder = [(0, root, None)]
        d = {None: 0}
        while postorder:
            seen, node, parent = postorder.pop()
            if not node: continue
            if not seen:
                postorder.append((1, node, parent))
                postorder.append((0, node.right, node.val))
                postorder.append((0, node.left, node.val))
            else:
                if node.val == parent:
                    d[node] = max(d[node.left], d[node.right]) + 1
                else:
                    d[node] = 0
                currMax = max(currMax, d[node.left] + d[node.right])
        return currMax
     

File: LeetCode/687_LongestUnivaluePath.go
// https://leetcode.cn/problems/longest-univalue-path/
// 2022年09月02日 07:46:02

// v1 recursive DFS
func longestUnivaluePath(root *TreeNode) int {
    currMax:=0
    var recDFS func(node *TreeNode) int
    recDFS = func(node *TreeNode) int {
        if node==nil {return 0}
        leftPath:=recDFS(node.Left)
        rightPath:=recDFS(node.Right)
        if node.Left!=nil && node.Val==node.Left.Val{
            leftPath++
        }else{
            leftPath=0
        }        
        if node.Right!=nil && node.Val==node.Right.Val{
            rightPath++
        }else{
            rightPath=0
        }
        if leftPath+rightPath>currMax{
            currMax=leftPath+rightPath
        }
        if leftPath>rightPath{
            return leftPath
        }
        return rightPath
    }
    recDFS(root)
    return currMax
}

// v1.1 recursive without local variable
func longestUnivaluePath(root *TreeNode) int {
    var recDFS func(node *TreeNode) (int,int)
    recDFS = func(node *TreeNode) (int,int) {
        if node==nil {return 0,0}
        maxL,pathL:=recDFS(node.Left)
        maxR,pathR:=recDFS(node.Right)
        currMax:=max(maxL,maxR)
        if node.Left!=nil && node.Val==node.Left.Val{
            pathL++
        }else{
            pathL=0
        }        
        if node.Right!=nil && node.Val==node.Right.Val{
            pathR++
        }else{
            pathR=0
        }
        currMax=max(currMax,pathL+pathR)
        return currMax, max(pathL,pathR)
    }
    ret,_:=recDFS(root)
    return ret
}
func max(a,b int) int{
    if a>b {return a}
    return b
}



// v2 iterative post-order DFS
// no idea, copied

func longestUnivaluePath(root *TreeNode) int {
    currMax:=0
    mixedType:=[]interface{} {0,root,nil}
    postorder := []interface{}
    d := make(map[*TreeNode]int)
    for len(postorder)>0{
        seen,node,parent := postorder[len(postorder)-1]
        postorder=postorder[:len(postorder)-1]
        if node==nil{
            continue
        }
        if seen==0{
            postorder=append(postorder,[]int{1,node,parent})
            postorder=append(postorder,[]int{0,node.Left,node.Val})
            postorder=append(postorder,[]int{0,node.Right,node.Val})
        }else{
            if node.Val==parent{
                d[node]:=max(d[node.Left],d[node.Right])+1
            }else{
                d[node]:=0
            }
            currMax=max(currMax,d[node.Left]+d[node.Right])
        }
    }
    return currMax
}
func max(a,b int)int{
    if a>b {return a}
    return b
}
